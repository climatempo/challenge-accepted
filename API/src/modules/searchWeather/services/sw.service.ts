import { Injectable } from '@nestjs/common';
import { LocaleDto } from '../dtos/locale.dto';
import { SearchLocaleDto } from '../dtos/searchLocale.dto';
import { SearchConditionDto } from '../dtos/searchCondition.dto';
import { ConditionDto } from '../dtos/condition.dto';
import { LocaleRepository } from '../repositories/locale.repository';
import { WeatherRepository } from '../repositories/weather.repository';
import { NotFoundException } from '../exceptions/notFound.exception';

/**
 * Classes de serviço são responsável pela regra de negócio do módulo
 */
@Injectable()
export class AppService {
  constructor(
    private readonly localeRepository: LocaleRepository,
    private readonly weatherRepository: WeatherRepository
  ) {}

  /**
   * Função responsável por processar as regras de negócio e comunicar com o repositório de dados
   *
   * @param dto //dados do usuário encapsulado e já validados
   * @returns LocaleDto //retorna um Data Transport Object contendo a resposta da pesquisa
   */
  public getWeatherLocation(dto: SearchLocaleDto): LocaleDto {
    let localeFound: LocaleDto;
    localeFound = this.localeRepository.getLocaleByName(
      dto.location.toLowerCase()
    );
    if (!localeFound) {
      throw new NotFoundException({}, 'A busca não encontrou um resultado');
    }

    return localeFound;
  }

  /**
   * Função responsável por processar as regras de negócio e comunicar com o repositório de dados
   *
   * @param dto //dados do usuário encapsulado e já validados
   * @returns ConditionDto //retorna um Data Transport Object contendo a resposta da pesquisa
   */
  public getConditionByLocale(dto: SearchConditionDto): ConditionDto {
    let conditionFound: ConditionDto;
    conditionFound = this.weatherRepository.getConditionByLocaleId(
      parseInt(dto.id)
    );
    if (!conditionFound) {
      throw new NotFoundException(
        { id: dto.id },
        'A busca não encontrou um resultado'
      );
    }

    return conditionFound;
  }
}
