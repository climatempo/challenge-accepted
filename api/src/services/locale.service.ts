import { Injectable } from '@nestjs/common';
import slugify from 'slug';

import locales from '../../../base/locales.json';
import { Locale } from '../models/locale.model';

@Injectable()
export class LocaleService {
    public async filterLocales(searchText: string): Promise<Locale[]> {
        const text = this.clearText(searchText);

        return locales.filter(l => this.clearText(l.name).includes(text));
    }

    private clearText(text: string): string {
        return slugify(text, {
            lower: true,
            replacement: '',
        });
    }
}
