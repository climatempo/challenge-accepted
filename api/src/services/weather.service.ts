import { Injectable } from '@nestjs/common';

import weather from '../../../base/weather.json';
import { Weather } from '../models/weather.model.js';

@Injectable()
export class WeatherService {
    public async getByLocaleId(localeId: number): Promise<Weather[]> {
        const weatherData = weather.find(w => w.locale.id === localeId);

        if (!weatherData) {
            return null;
        }

        return weatherData.weather.map(w => {
            return {
                ...w,
                rain: {
                    precipitation: +w.rain.precipitation, // alguns dados est√£o como string, padronizando o retorno da api
                    probability: w.rain.probability,
                },
            };
        });
    }
}
