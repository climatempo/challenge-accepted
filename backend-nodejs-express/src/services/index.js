import path from 'path';
import fs from 'fs';

export default (app) => {
  try {
    const normalizedPath = path.join(__dirname);
    fs.readdirSync(normalizedPath).forEach(curr => {
      const pathToLoad = `${normalizedPath}/${curr}`;
      if (fs.statSync(pathToLoad).isDirectory()) {
        fs.readdirSync(pathToLoad).forEach(file => {
          if (file === 'index.js') require(`${pathToLoad}/${file}`).default(app);
        });
      }
    });
  } catch (err) {
    global.console.error(err);
  }
};
