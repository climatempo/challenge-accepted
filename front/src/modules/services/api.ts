import axios from "axios";

const baseURL = process.env.API_ENDPOINT || "http://back:3000";

const api = axios.create({
  baseURL,
});

export default api;
