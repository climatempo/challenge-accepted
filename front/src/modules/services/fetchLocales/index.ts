import { Dispatch, SetStateAction } from "react";
import removeSpecialChars from "../../removeSpecialChars";
import { Locale } from "../../types/data";
import { locales } from "../mock";
// import api from "../api";

async function fetchLocales(
  setLocales: Dispatch<SetStateAction<Locale[] | null>>,
  searchValue: string,
  setIsLoading?: Dispatch<SetStateAction<boolean>>
) {
  if (setIsLoading) setIsLoading(true);

  const formattedSearchValue = removeSpecialChars(searchValue);

  // const { data } = await api.get("/locales", {
  //   params: {
  //     query: searchValue ? searchValue : "",
  //   }
  // })

  const filteredLocales = locales.filter(({ name, state }) => {
    const formattedName = removeSpecialChars(`${name} - ${state}`);
    return formattedName.includes(formattedSearchValue);
  });

  setTimeout(() => {
    if (setIsLoading) setIsLoading(false);
    if (filteredLocales.length) setLocales(filteredLocales);
  }, 1000);
}

export default fetchLocales;
