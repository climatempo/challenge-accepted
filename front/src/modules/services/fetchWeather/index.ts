import { Dispatch, SetStateAction } from "react";
import { Weather } from "../../types/data";
import { weathers } from "../mock";
// import api from "../api";

async function fetchWeather(
  setWeather: Dispatch<SetStateAction<Weather | null>>,
  id: string,
  setIsLoading?: Dispatch<SetStateAction<boolean>>
) {
  // const { data } = await api.get("/weather", {
  //   params: {
  //     id,
  //   }
  // })

  const weather = weathers.find(({ locale }) => String(locale.id) === id);

  setTimeout(() => {
    if (weather) setWeather(weather);
    if (setIsLoading) setIsLoading(false);
  }, 2000);
}

export default fetchWeather;
