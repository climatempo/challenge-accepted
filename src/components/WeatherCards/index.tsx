import WeatherCardContainer from "../WeatherCard/container";
import Wrapper from "./styles";
import { Props } from "./types";

function WeatherCards({ weather }: Props) {
  return (
    <Wrapper>
      {weather.weather.map((item, index) => (
        <WeatherCardContainer data={item} key={index} />
      ))}
    </Wrapper>
  );
}

export default WeatherCards;
