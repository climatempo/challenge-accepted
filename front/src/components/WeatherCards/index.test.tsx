import { render } from "@testing-library/react";
import WeatherCards from ".";
import { weathers } from "../../modules/services/mock";

describe("WeatherCards", () => {
  it("renders correctly", () => {
    const { container } = render(
      <WeatherCards
        weather={weathers[0]}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
