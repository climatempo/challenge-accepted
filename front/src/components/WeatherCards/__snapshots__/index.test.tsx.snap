// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCards renders correctly 1`] = `
.c5 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  width: min(100%,30rem);
  box-shadow: 0 1rem 1rem 0.2rem rgba(0,0,0,0.2);
  margin: 2rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
  padding: 20px;
}

.c3 {
  font-size: 1.6rem;
}

.c4 {
  font-size: 1.4rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  padding: 20px 0;
  background: #e3e3e3;
}

.c7 {
  display: grid;
  grid-template-columns: repeat(2,auto);
  grid-gap: 20px 40px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c9 {
  width: 2rem;
}

.c10 {
  font-size: 2rem;
  color: #2086c3;
}

.c11 {
  font-size: 2rem;
  color: #cd4b4b;
}

.c12 {
  font-size: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding-bottom: 20px;
}

@media (max-width:600px) {
  .c1 {
    margin-left: 0;
    margin-right: 0;
  }
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          31/01/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com muitas nuvens durante o dia. Períodos de nublado, com chuva a qualquer hora.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              82 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              68 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.79 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          01/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com muitas nuvens durante o dia. Períodos de nublado, com chuva a qualquer hora.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              86 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              70 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.39 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          02/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com algumas nuvens. Chove rápido durante o dia e à noite.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              88 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              72 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.59 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          03/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com algumas nuvens. Chove rápido durante o dia e à noite.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              90 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              72 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.63 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          04/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol e aumento de nuvens de manhã. Pancadas de chuva à tarde e à noite.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              90 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              73 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.75 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              67%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          05/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com algumas nuvens. Chove rápido durante o dia e à noite.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              91 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              72 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.31 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="c1"
    >
      <section
        class="c2"
      >
        <h3
          class="c3"
        >
          06/02/2017
        </h3>
        <span
          class=" c4"
        >
          Sol com algumas nuvens. Chove rápido durante o dia e à noite.
        </span>
      </section>
      <div
        class="c5 c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowup.png"
            />
            <span
              class=" c10"
              color="#2086c3"
            >
              86 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/arrowdown.png"
            />
            <span
              class=" c11"
              color="#cd4b4b"
            >
              77 °F
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/raindrop.png"
            />
            <span
              class=" c12"
            >
              0.43 inch
            </span>
          </li>
          <li
            class="c8"
          >
            <img
              class=" c9"
              src="/icons/umbrella.png"
            />
            <span
              class=" c12"
            >
              60%
            </span>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>
`;
